LoongArch-ELF-ABI-EN.adoc

```diff
@@ -0,0 +1,206 @@
+|$r22     |$fp          |Frame pointer       |Yes
+|$r23-$r31|$s0-$s8      |Subroutine register variables   |Yes
+|====================================================================
+
+.Floating-point Register Convention
+[width="80%",options="header",cols="2m,2m,^5,^3"]
+|====================================================================
+|Name     |Alias        |Meaning            |Preserved across calls
+|$f0-$f7  |$fa0-$fa7    |Argument registers |No
+|$f0-$f1  |$fv0-$fv7    |Return value       |No
+|$f8-$f23 |$ft0-$ft15   |Temp registers     |No
+|$f24-$f31|$fs0-$fs7    |Subroutine register variables   |Yes
+|====================================================================
+
+
+== Type Size and Alignment
+
+.LP64 ABI
+[width="80%",options="header",cols="2m,^5,^3"]
+|====================================================================
+|Scalar type               |Size |Alignment
+|unsigned/signed char      |1    |1
+|unsigned/signed short     |2    |2
+|unsigned/signed int       |4    |4
+|unsigned/signed long      |8    |8
+|unsigned/signed long long |8    |8
+|pointer                   |8    |8
+|float                     |4    |4
+|double                    |8    |8
+|long double               |16   |16
+|====================================================================
+
+== ELF Object Files
+
+e_machine: LoongArch (258)
+
+e_flags: LP32 (0x1) LP64 (0x3)
+
+== Relocations
+.ELF Relocation types
+[width="100%",options="header",cols="^1,^1m,^10,10"]
+|====================================================================
+|Enum    |ELF reloc type          |Usage      |Detail
+|0       |R_LARCH_NONE            |           |
+
+|1	 |R_LARCH_32	          |Runtime address resolving|
+`*(int32_t *) PC = RtAddr + A`
+
+|2	 |R_LARCH_64	          |Runtime address resolving|
+`*(int64_t *) PC = RtAddr + A`
+
+|3       |R_LARCH_RELATIVE        |Runtime fixup for load-address|
+`*(void **) PC = B + A`
+
+|4       |R_LARCH_COPY            |Runtime memory copy in executable|
+`memcpy (PC, RtAddr, sizeof (sym))`
+
+|5       |R_LARCH_JUMP_SLOT       |Runtime PLT supporting|
+_implementation-defined_
+
+|6       |R_LARCH_TLS_DTPMOD32    |Runtime relocation for TLS-GD|
+`*(int32_t *) PC = ID of module defining sym`
+
+|7       |R_LARCH_TLS_DTPMOD64    |Runtime relocation for TLS-GD|
+`*(int64_t *) PC = ID of module defining sym`
+
+|8       |R_LARCH_TLS_DTPREL32    |Runtime relocation for TLS-GD|
+`*(int32_t *) PC = DTV-relative offset for sym`
+
+|9       |R_LARCH_TLS_DTPREL64    |Runtime relocation for TLS-GD|
+`*(int64_t *) PC = DTV-relative offset for sym`
+
+|10      |R_LARCH_TLS_TPREL32     |Runtime relocation for TLE-IE|
+`*(int32_t *) PC = T`
+
+|11      |R_LARCH_TLS_TPREL64     |Runtime relocation for TLE-IE|
+`*(int64_t *) PC = T`
+
+|12      |R_LARCH_IRELATIVE       |Runtime local indirect function resolving|
+`*(void **) PC = (((void *)(*)()) (B + A)) ()`
+
+4+|...
+
+|20      |R_LARCH_MARK_LA            |Mark la.abs|
```

Comment:
What is la.abs?

Reply:
> What is la.abs?

A macro instruction including fixed 4 machine instructions to load 64-bit address.